src/Control/Algebra.hs

```diff
@@ -90,20 +91,19 @@ class Monad m => Algebra sig m | m -> sig where
     -> ctx ()          -- ^ The initial state.
     -> m (ctx a)       -- ^ The interpretation of the effect in @m@.
 
--- | Compose and thread a pair of handlers and input state through the algebra for some underlying signature.
+-- | Thread a composed handler and input state through the algebra for some underlying signature.
 --
 -- @since 1.1.0.0
 thread
   :: ( Functor ctx1
      , Functor ctx2
      , Algebra sig m
      )
-  => Handler ctx1 n m
-  -> Handler ctx2 o n
-  -> sig o a
+  => Handler (Compose ctx1 ctx2) n m
+  -> sig n a
   -> ctx1 (ctx2 ())
   -> m (ctx1 (ctx2 a))
-thread hdl1 hdl2 sig = fmap getCompose . alg (fmap Compose . hdl1 . fmap hdl2 . getCompose) sig . Compose
+thread hdl sig = fmap getCompose . alg hdl sig . Compose
```

Comment:
Passing a composed handler to `thread` is more flexible: if, for some reason, `fmap` is particularly expensive for some contexts, handlers are free to do something else.

Reply:
This change also motivated the addition of `~<~` and the (public) `Control.Algebra.Handler` module.